File: Files/Docs/Мультиметры управление etc/ADL-GPIB/Samples/BCB/Async Read Write/Unit1.cpp
/********************************************************************
*   Example: Read/Write
*
*   Description: Perform the asynchronous read/write operation.
*
*   ADLINK Inc. 2006
********************************************************************/	
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TForm1::btnRWClick(TObject *Sender)
{
  //Declare variables
  AnsiString wrtbuf, Reading;
  char wr_str[100],rdbuf[100];
  int dev;

  //Open and intialize an GPIB instrument
  dev = ibdev(cmbGPIB->ItemIndex, cmbInst->ItemIndex+1, 0, T1s, 1, 0);
  if (ibsta & ERR)
  {
    ShowMessage("Error in initializing the GPIB instrument.");
  }

  //clear a specific device
  ibclr(dev);
  if (ibsta & ERR)
  {
    ShowMessage("Error in clearing the GPIB device.");
    ibonl(dev, 0);
    return;
  }

  wrtbuf = editWrite->Text;
  //Write a string command to a GPIB instrument asynchronously using the ibwrta() command
  ibwrta(dev, wrtbuf.c_str(), wrtbuf.Length());
  if (ibsta & ERR)
  {
    ShowMessage("Error in writing the string command to the GPIB instrument.");
    ibonl(dev, 0);
    return;
  }

  //Wait for the completion of asynchronous write operation
  ibwait(dev, CMPL);
  if (ibsta & ERR)
  {
    ShowMessage("Writing the string command to the GPIB instrument timeout.");
    ibonl(dev, 0);
    return;
  }

  //Read the response string from the GPIB instrument asynchronously using the ibrda() command
  ibrda(dev, rdbuf, 100);
  if (ibsta & ERR)
  {
    ShowMessage("Error in reading the response string from the GPIB instrument.");
    ibonl(dev, 0);
    return;
  }

  //Wait for the completion of asynchronous read operation
  ibwait(dev, CMPL);
  if (ibsta & ERR)
  {
    ShowMessage("Reading the string command to the GPIB instrument timeout.");
    ibonl(dev, 0);
    return;
  }

  rdbuf[ibcnt]=0;

  memoRead->Text = rdbuf;

  //Offline the GPIB interface card
  ibonl(dev, 0);
  if (ibsta & ERR)
  {
    ShowMessage("Error in offline the GPIB interface card.");
  }

}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  memoRead->Text = "";

}
//---------------------------------------------------------------------------

File: Files/SRC/agilent 3458a/3548a/Unit1.pas
(********************************************************************
*   Example: Query Identification
*
*   Description: Perform the asynchronous read/write operation.
*
*   ADLINK Inc. 2006
********************************************************************)	
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, gpib_user, Adgpib;

type
  TForm1 = class(TForm)
    Label1: TLabel;
    cmbGPIB: TComboBox;
    Label2: TLabel;
    cmbInst: TComboBox;
    Button1: TButton;
    Label3: TLabel;
    editWrite: TEdit;
    Label4: TLabel;
    memoRead: TMemo;
    Memo1: TMemo;
    procedure Button1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

  (*488.2 GPIB global status variables.*)
  ibsta  : Integer;
  iberr  : Integer;
  ibcnt  : Integer;
  ibcntl : Integer;

implementation

{$R *.DFM}

procedure TForm1.Button1Click(Sender: TObject);
var
  //Declare variables
  wrtbuf, rdbuf: packed array[0..99] of char  ;
  dev: integer;
  sta : integer;
  Procedure msg(mss:string);
  begin
      memoRead.lines.add(mss+' ibsta='+IntToStr(ibsta)+' err='+IntToStr(iberr)+' ibcntl='+IntToStr(ibcntl));
  end;
begin
  //Open and intialize an GPIB instrument
  dev := ibdev(cmbGPIB.ItemIndex, cmbInst.ItemIndex+1, 0, T1s, 1, 0);
  gpib_get_globals(@ibsta, @iberr, @ibcnt, @ibcntl);
  msg('ibdev');
  //clear the specific GPIB instrument
  ibclr(dev);
  gpib_get_globals(@ibsta, @iberr, @ibcnt, @ibcntl);
  msg('ibclr');

  strcopy(wrtbuf, pchar(editWrite.Text));
  //Write a string command to a GPIB instrument asynchronously using the ibwrta() command
  ibwrta(dev, @wrtbuf, strlen(wrtbuf));
  gpib_get_globals(@ibsta, @iberr, @ibcnt, @ibcntl);
  msg('ibwrt');

  //Wait for the completion of asynchronous write operation
  ibwait(dev, CMPL);
  gpib_get_globals(@ibsta, @iberr, @ibcnt, @ibcntl);
//  if (ibsta AND ERR) <> 0 THEN
  msg('ibwait');

  //Read the response string from the GPIB instrument asynchronously using the ibrda() command
  ibrda(dev, @rdbuf, 100);
  msg('ibrda');

  //Wait for the completion of asynchronous read operation
//  sleep(1000);
  ibwait(dev, CMPL);
  msg('ibwait');
  rdbuf[ibcnt] := chr(0);

  memoRead.lines.add(rdbuf);

  ibonl(dev, 0);
  gpib_get_globals(@ibsta, @iberr, @ibcnt, @ibcntl);
  msg('ibonl');

end;

procedure TForm1.FormCreate(Sender: TObject);
begin
  cmbGPIB.ItemIndex := 0;
  cmbInst.ItemIndex := 21;
  memoRead.Text := '';

end;

end.
